src/rt/region/region_rc.h

```diff
@@ -0,0 +1,391 @@
+   * items will first query the freelist to see if a hole can be reused,
+   * otherwise items are bump allocated.
+   *
+   * To maintain an internal freelist with no additional space requirements, the
+   * item `T` must be at least 1 machine word in size.
+   */
+  template<class T, class Alloc>
+  class RegionVector
+  {
+    static constexpr size_t ITEM_COUNT = 32;
+    static_assert(
+      snmalloc::bits::next_pow2_const(ITEM_COUNT) == ITEM_COUNT,
+      "Should be power of 2 for alignment.");
+
+    static constexpr size_t BLOCK_COUNT = ITEM_COUNT - 1;
+    static constexpr uintptr_t EMPTY_MASK = 1 << 0;
+
+    struct alignas(ITEM_COUNT * sizeof(T)) Block
+    {
+      T prev;
+      T data[BLOCK_COUNT];
+    };
+
+  private:
+    // Dummy block to effectively allow pointer arithmetic on nullptr
+    // which is undefined behaviour.  So we statically allocate a block
+    // to represent the end of the vec.
+    inline static Block null_block{};
+
+    // Index of the full dummy block
+    // Due to pointer arithmetic with nullptr being undefined behaviour
+    // we use a statically allocated null block.
+    static constexpr T* null_index = &(null_block.data[BLOCK_COUNT - 1]);
+
+    /// Mask to access the index component of the pointer to a block.
+    static constexpr uintptr_t INDEX_MASK = (ITEM_COUNT - 1) * sizeof(T);
+
+    /// Pointer into a block.  As the blocks are strongly aligned
+    /// the bits 9-3 represent the element in the block, with 0 being
+    /// a pointer to the `prev` pointer, and implying the empty block.
+    T* index;
+
+    // Used to thread a freelist pointer through the vec.
+    T* next_free;
+
+    /// Takes an index and returns the pointer to the Block
+    static Block* get_block(T* ptr)
+    {
+      return snmalloc::pointer_align_down<sizeof(Block), Block>(ptr);
+    }
+
+    /// Checks if an index into a block means the block is empty.
+    static bool is_empty(T* ptr)
+    {
+      return ((uintptr_t)ptr & INDEX_MASK) == 0;
+    }
+
+    /// Checks if an index into a block means the block has no space.
+    static bool is_full(T* index)
+    {
+      return ((uintptr_t)index & INDEX_MASK) == INDEX_MASK;
+    }
+
+  public:
+    RegionVector<T, Alloc>() : index(null_index), next_free(nullptr)
+    {
+      static_assert(
+        sizeof(*this) == sizeof(void*) * 2,
+        "Stack should contain only the index and freelist pointer");
+    }
+
+    /// Deallocate the linked blocks for this vec.
+    void dealloc(Alloc& alloc)
+    {
+      auto local_block = get_block(index);
+      while (local_block != &null_block)
+      {
+        auto prev_ptr = *(T**)&(local_block->prev);
+        auto prev = get_block(prev_ptr);
+        alloc.template dealloc<sizeof(Block)>(local_block);
+        local_block = prev;
+      }
+      index = null_index;
+    }
+
+    ALWAYSINLINE void remove(T* index)
+    {
+      *(uintptr_t*)index = ((uintptr_t)next_free | EMPTY_MASK);
```

Comment:
Assert that the `next_free` does not have the bit set, or perhaps better `index`

Reply:
Yeah, there is definitely an implicit assumption that `T` starts with a pointer. I think a constraint here is a good idea.

Reply:
I've reworked this now to be less generic and use unions to specify that the vector might have holes in it (fd32d6).